sdk/core/core/inc/az_platform.h

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef _az_PLATFORM_H
+#define _az_PLATFORM_H
+
+#include <az_result.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <_az_cfg_prefix.h>
+
+/**
+ * @brief az_platform_clock_msec gets platform clock in milliseconds.
+ * The moment of time where clock starts is undefined, but if this function is getting called twice
+ * with one second interval, the difference between the values returned should be equal to 1000.
+ * @return Platform clock in milliseconds.
+ */
+AZ_NODISCARD int64_t az_platform_clock_msec();
+
+/**
+ * @brief az_platform_sleep_msec tells platform to sleep for a given number of milliseconds.
+ * @param milliseconds number of milliseconds to sleep.
+ * @remarks For passing \p milliseconds with negative values and 0, behavior is undefined.
+ */
+void az_platform_sleep_msec(int32_t milliseconds);
+
+/**
+ * @brief az_platform_atomic_compare_exchange Conditionally exchanges values of two pointers in a
+ * thread-safe manner.
+ *
+ * @param obj a pointer to a pointer that needs to be changed.
+ * @param expected an expected value of a value that \p obj is pointing to, prior to exchange.
+ * @param desired a value to assign to the value that is pointed by \p obj, if its value equals to
+ * \p expected.
+ *
+ * @return `true` if previous value of \p obj matches the \p expected.
+ *
+ * @remarks If `*obj` equals to \p expected, its value is being overwritten with \p desired, and no
+ * operation otherwise. Returns `true` if the value was overwritten, `false` otherwise.
+ */
+AZ_NODISCARD bool az_platform_atomic_compare_exchange(
+    void* volatile* obj,
```

Comment:
`volatile*` - What does this mean?

Reply:
So basically when you see a pointer, whatever is on the left side of `*` characterizes what it is pointing to. On the right - characterizes the pointer itself.
Remember this example?:
It is easier when you read pointers from right to left:
 `int const*` - non-constant pointer to constant int
`int* const` - constant pointer to non-constant int.
`int const* const` - constant pointer to constant int
`int*` - non-constant pointer to non-constant int.
... and so on, with pointers-to-pointers, it can take really monstrous shapes:
`int* const*** const* p` - p is a non-constant pointer to constant pointer to non-constant pointer to non-constant pointer to constant pointer to non-constant int.

A `const` is a modifier, and so is `volatile` (together, they are called "CV-modifiers"). Everywhere where you can use `const`, you can use `volatile`. You can use them both at the same time. Or just one. Or none of them.

And `void*` is "just" a pointer, to an unknown type. So, before you read or write to the value pointed by such pointer (`*p = 3`), you would need to cast it to a pointer of some type, other than `void`: *(int*)p = 3; *(double*)p = 2.0.

so, to read this, `void* volatile*` is a  (non-constant) pointer to a volatile pointer (to ???).